carapace-server/src/main/java/org/carapaceproxy/configstore/HerdDBConfigurationStore.java

```diff
@@ -475,6 +494,21 @@ public void saveCertificate(CertificateData cert) {
         }
     }
 
+    @Override
+    public void removeCertificate(final String certId) {
+        try {
+            final var connection = datasource.getConnection();
+            final var preparedStatement = connection.prepareStatement(REMOVE_DIGITAL_CERTIFICATES_TABLE);
```

Comment:
why not opening both in the twr? if the `prepareStatement` call fails the connection is not closed

Reply:
The try-with-resources is in the line below: from Java 9 it should be an alternatively valid syntax

Reply:
It's valid syntax but I think the compiled code does not have the same meaning. By declaring the resources outside the twr, if the call to `connection#prepareStatement` fails then `connection#close` is not invoked.

PS. actually tested out of curiosity and I can confirm this. With this syntax the 2 resources are declared and then the 2 nested try are created. By declaring both inside the twr you end up with the first declaration and the first try block, which wraps the second declaration and the second try block

Reply:
Cool, didn't know that, I totally missed the point of that addition, thank you!

I also tried it like that:

```java
public class Test {

    public interface Common extends AutoCloseable {
        @Override
        default void close() {
            System.out.println(this.getClass() + " closed correctly");
        }

        default Successful getSuccessful() {
            return new Successful();
        }
        
        default Failure getFailure() {
            return new Failure();
        }
    }

    public static class Successful implements Common {
        public Successful() {
            System.out.println(this.getClass() + " created correctly");
        }
    }

    public static class Failure implements Common {
        public Failure() {
            System.out.println(this.getClass() + " created");
            throw new IllegalStateException("Faiulre");
        }
    }

    public static void classic() {
        try (final var succ = new Successful(); final var fail = succ.getFailure()) {
            System.out.println("Try block");
        }
    }

    public static void java9() {
        final var succ = new Successful();
        try (final var fail = succ.getFailure()) {
            System.out.println("Try block");
        }
    }

    public static void likeIDid() {
        final var succ = new Successful();
        final var fail = succ.getFailure();
        try (succ; fail) {
            System.out.println("Try block");
        }
    }

    public static void main(final String[] args) {
        System.out.println("classic...");
        try {
            classic();
        } catch (Exception ignored) { }
        System.out.println("Java 9...");
        try {
            java9();
        } catch (Exception ignored) { }
        System.out.println("Like I did...");
        try {
            likeIDid();
        } catch (Exception ignored) { }
    }
}
```

which is compiled like that:

```java
//
// Source code recreated from a .class file by IntelliJ IDEA
// (powered by FernFlower decompiler)
//

public class Test {
    public Test() {
    }

    public static void classic() {
        Successful var0 = new Successful();

        try {
            Failure var1 = var0.getFailure();

            try {
                System.out.println("Try block");
            } catch (Throwable var6) {
                if (var1 != null) {
                    try {
                        var1.close();
                    } catch (Throwable var5) {
                        var6.addSuppressed(var5);
                    }
                }

                throw var6;
            }

            if (var1 != null) {
                var1.close();
            }
        } catch (Throwable var7) {
            try {
                var0.close();
            } catch (Throwable var4) {
                var7.addSuppressed(var4);
            }

            throw var7;
        }

        var0.close();
    }

    public static void java9() {
        Successful var0 = new Successful();
        Failure var1 = var0.getFailure();

        try {
            System.out.println("Try block");
        } catch (Throwable var5) {
            if (var1 != null) {
                try {
                    var1.close();
                } catch (Throwable var4) {
                    var5.addSuppressed(var4);
                }
            }

            throw var5;
        }

        if (var1 != null) {
            var1.close();
        }

    }

    public static void likeIDid() {
        Successful var0 = new Successful();
        Failure var1 = var0.getFailure();
        Successful var2 = var0;

        try {
            Failure var3 = var1;

            try {
                System.out.println("Try block");
            } catch (Throwable var8) {
                if (var1 != null) {
                    try {
                        var3.close();
                    } catch (Throwable var7) {
                        var8.addSuppressed(var7);
                    }
                }

                throw var8;
            }

            if (var1 != null) {
                var1.close();
            }
        } catch (Throwable var9) {
            if (var0 != null) {
                try {
                    var2.close();
                } catch (Throwable var6) {
                    var9.addSuppressed(var6);
                }
            }

            throw var9;
        }

        if (var0 != null) {
            var0.close();
        }

    }

    public static void main(String[] var0) {
        System.out.println("classic...");

        try {
            classic();
        } catch (Exception var4) {
        }

        System.out.println("Java 9...");

        try {
            java9();
        } catch (Exception var3) {
        }

        System.out.println("Like I did...");

        try {
            likeIDid();
        } catch (Exception var2) {
        }

    }

    public static class Successful implements Common {
        public Successful() {
            System.out.println(String.valueOf(this.getClass()) + " created correctly");
        }
    }

    public static class Failure implements Common {
        public Failure() {
            System.out.println(String.valueOf(this.getClass()) + " created");
            throw new IllegalStateException("Faiulre");
        }
    }

    public interface Common extends AutoCloseable {
        default void close() {
            System.out.println(String.valueOf(this.getClass()) + " closed correctly");
        }

        default Successful getSuccessful() {
            return new Successful();
        }

        default Failure getFailure() {
            return new Failure();
        }
    }
}
```

and in fact, if run, gives the behavior that you were hinting:

```
❯ java Test                                                                                                          
classic...
class Test$Successful created correctly
class Test$Failure created
class Test$Successful closed correctly
Java 9...
class Test$Successful created correctly
class Test$Failure created
Like I did...
class Test$Successful created correctly
class Test$Failure created
```

Thank you for pointing it out, I'll fix it